src/box/space.c

```diff
@@ -325,6 +326,123 @@ space_unpin_defaults(struct space *space)
 	}
 }
 
+/**
+ * Set all the events associated with space: on_replace and before_replace,
+ * bound by name and by id. If argument on_recovery is true, recovery events
+ * are set instead of regular ones.
+ */
+static void
+space_set_events(struct space *space, bool on_recovery)
+{
+	assert(space->def != NULL);
+	/*
+	 * Space events - on_replace and before_replace events. All event names
+	 * must be declared in the same order.
+	 */
+	struct space_event *space_events[] = {
+		&space->on_replace_event,
+		&space->before_replace_event,
+	};
+	/* Format strings for names of events associated with space by id. */
+	static const char * const regular_event_by_id_fmt[] = {
+		"box.space[%u].on_replace",
+		"box.space[%u].before_replace",
+	};
+	static const char * const recovery_event_by_id_fmt[] = {
+		"box.space[%u].on_recovery_replace",
+		"box.space[%u].before_recovery_replace",
+	};
+	/* Format strings for names of events associated with space by name. */
+	static const char * const regular_event_by_name_fmt[] = {
+		"box.space.%s.on_replace",
+		"box.space.%s.before_replace",
+	};
+	static const char * const recovery_event_by_name_fmt[] = {
+		"box.space.%s.on_recovery_replace",
+		"box.space.%s.before_recovery_replace",
+	};
+	assert(lengthof(space_events) == lengthof(regular_event_by_id_fmt));
+	assert(lengthof(space_events) == lengthof(recovery_event_by_id_fmt));
+	assert(lengthof(space_events) == lengthof(regular_event_by_name_fmt));
+	assert(lengthof(space_events) == lengthof(recovery_event_by_name_fmt));
+
+	const char * const *event_by_id_fmt = NULL;
+	const char * const *event_by_name_fmt = NULL;
+	space->run_recovery_triggers = on_recovery;
+	if (on_recovery) {
+		event_by_id_fmt = recovery_event_by_id_fmt;
+		event_by_name_fmt = recovery_event_by_name_fmt;
+	} else {
+		event_by_id_fmt = regular_event_by_id_fmt;
+		event_by_name_fmt = regular_event_by_name_fmt;
+	}
+	/*
+	 * Allocate buffer for event names. We make an assumption that all
+	 * format strings for event names associated by space name are limited
+	 * by the constant below. This assumption is tested in the loop at the
+	 * end of the function.
+	 * Another assumption is that names of events associated with space by
+	 * id are shorted than TT_STATIC_BUF_LEN since the string representation
+	 * of uint32_t is limited by 10 characters and TT_STATIC_BUF_LEN is
+	 * much bigger.
+	 * So the snippet chooses between tt_static_buf and malloc depending on
+	 * maximal length of the name of event associated by the space name.
+	 */
+	const size_t event_by_name_fmt_max_len = 40;
+	size_t buf_size = event_by_name_fmt_max_len +
+		strlen(space->def->name) + 1;
+	buf_size = MAX(buf_size, TT_STATIC_BUF_LEN);
+	char *event_name = NULL;
+	if (buf_size > TT_STATIC_BUF_LEN)
+		event_name = xmalloc(buf_size);
```

Comment:
It should be `xregion_alloc` and `region_truncated`.

Reply:
Done.